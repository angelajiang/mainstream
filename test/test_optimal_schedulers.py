import os
import shutil
import subprocess
import pytest

@pytest.mark.unit
@pytest.mark.parametrize("scheduler_type", ["hifi", "stems"])
def test_optimality(scheduler_type):
    output_dir = os.path.join("test", "tmp", scheduler_type)
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    schedules_dir = os.path.join(output_dir, "schedules")
    if not os.path.exists(schedules_dir):
        os.makedirs(schedules_dir)
    other_file = os.path.join(schedules_dir, scheduler_type + ".sim.debug.v0")
    exhaustive_file = os.path.join(schedules_dir, "exhaustive.sim.debug.v0")

    subprocess.check_call("test/exhaustive_vs_x_scheduler.sh " + scheduler_type, shell=True)

    other_f1s = []
    with open(other_file, "r") as f:
        for line in f:
            vals = line.split(',')
            metric = round(float(vals[1]), 2)
            other_f1s.append(metric)

    exhaustive_f1s = []
    with open(exhaustive_file, "r") as f:
        for line in f:
            vals = line.split(',')
            metric = round(float(vals[1]), 2)
            exhaustive_f1s.append(metric)

    assert other_f1s == exhaustive_f1s

    shutil.rmtree(output_dir)


@pytest.mark.unit
@pytest.mark.parametrize("scheduler_type,num_apps_range", [
    ("hifi", 5),
    ("stems", 5),
    ("exhaustive", 3),
    ("stems_cpp", 5),
])
def test_regression(regtest, scheduler_type, num_apps_range):
    output_dir = os.path.join("test", "tmp", scheduler_type)
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    schedules_dir = os.path.join(output_dir, "schedules")
    if not os.path.exists(schedules_dir):
        os.makedirs(schedules_dir)
    other_file = os.path.join(schedules_dir, scheduler_type + ".sim.debug.v0")

    subprocess.check_call("test/run_scheduler_with_setups.sh {} {}".format(scheduler_type, num_apps_range), shell=True)

    with open(other_file) as f:
        for line in f:
            line = line.split(",")[:-1]
            regtest.write(",".join(line) + "\n")

    shutil.rmtree(output_dir)
