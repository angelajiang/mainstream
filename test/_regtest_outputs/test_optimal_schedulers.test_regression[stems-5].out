2b73dcb7.v1,2,0.070543,27,27,4,4,100
2017004d.v1,2,0.216463,27,27,3,4,100
bc3a2366.v1,2,0.217071,51,21,2,4,100
ed3df60a.v1,2,0.321578,63,63,4,4,100
d2e344c2.v1,2,0.201458,27,27,4,4,100
ae4508c7.v1,3,0.318509,45,51,45,4,4,4,100
3b96ef08.v1,3,0.142924,75,27,75,3,4,4,100
734b24d2.v1,3,0.19328,33,51,39,3,4,4,100
2a9ac25c.v1,3,0.126833,33,33,51,4,3,3,100
a9dc0986.v1,3,0.314313,51,39,33,4,4,3,100
59b6c67d.v1,4,0.339579,75,63,33,75,4,4,4,4,100
d9d27a0c.v1,4,0.20886,39,75,75,57,4,3,4,4,100
17d0cda0.v1,4,0.16885,51,57,57,57,4,4,3,3,100
2852935c.v1,4,0.355489,57,63,63,33,4,3,2,4,100
e7219e43.v1,4,0.324825,63,63,51,51,4,3,3,4,100
61094175.v1,5,0.116255,69,69,57,57,69,4,3,3,4,4,100
8d77f3d8.v1,5,0.278895,75,63,75,33,51,4,4,3,2,3,100
9d9c6291.v1,5,0.238785,75,33,75,51,75,4,4,3,2,3,100
aeaba1c7.v1,5,0.202643,57,33,63,69,69,3,2,4,4,3,100
f306b783.v1,5,0.373231,63,69,69,63,63,4,4,4,4,3,100
